File: backend/services/wallet_jetton_contract_verification.py
"""Immutable proof-checked jetton wallet/master relationship evidence."""

from __future__ import annotations

from datetime import datetime, timezone
import hashlib
import json
import re
from typing import Any, Callable

from pytoniq_core import Cell
from sqlalchemy import select
from sqlalchemy.orm import Session

from config import get_settings
from models import (
    WalletBalanceSnapshot,
    WalletIngestionRun,
    WalletJettonContractVerification,
)
from services.ton_address_identity import derive_ton_wallet_identity
from services.ton_liteclient_jetton_verifier import (
    TonLiteclientJettonVerificationFailure,
    verify_jetton_contract_relationship_live,
)


JETTON_CONTRACT_VERIFICATION_VERSION = "ton_jetton_contract_verification_v1"
JETTON_ASSET_IDENTITY_VERSION = "ton_jetton_asset_v1"


class WalletJettonContractVerificationNotFound(LookupError):
    """The run or exact persisted jetton snapshot relation was not found."""


class WalletJettonContractVerificationConflict(ValueError):
    """Persisted or live contract relationship evidence is incoherent."""


class WalletJettonContractVerificationFailure(RuntimeError):
    """The local proof verifier or immutable storage was unavailable."""


def verify_wallet_jetton_contract_relationship(
    run_id: int,
    jetton_wallet_account_canonical: str,
    jetton_master_account_canonical: str,
    session: Session,
    *,
    live_verifier: Callable[..., dict[str, Any]] = (
        verify_jetton_contract_relationship_live
    ),
) -> dict[str, Any]:
    """Verify and persist one exact selected-run jetton contract relation."""
    run = session.get(WalletIngestionRun, run_id)
    if run is None:
        raise WalletJettonContractVerificationNotFound(
            "Persisted wallet ingestion run not found."
        )
    network = _validated_run_network(run)
    wallet = _canonical_account(jetton_wallet_account_canonical, network)
    master = _canonical_account(jetton_master_account_canonical, network)
    if wallet != jetton_wallet_account_canonical or master != (
        jetton_master_account_canonical
    ):
        raise WalletJettonContractVerificationConflict(
            "Jetton verification addresses must be canonical raw TON addresses."
        )
    snapshot = _exact_snapshot(run_id, wallet, master, network, session)
    existing = session.scalar(
        select(WalletJettonContractVerification).where(
            WalletJettonContractVerification.run_id == run_id,
            WalletJettonContractVerification.jetton_wallet_account_canonical
            == wallet,
            WalletJettonContractVerification.jetton_master_account_canonical
            == master,
            WalletJettonContractVerification.contract_version
            == JETTON_CONTRACT_VERIFICATION_VERSION,
        )
    )
    if existing is not None:
        return _record_response(existing)

    settings = get_settings()
    try:
        verified = live_verifier(
            network=network,
            owner_account_canonical=run.wallet_address_canonical,
            jetton_wallet_account_canonical=wallet,
            jetton_master_account_canonical=master,
            trust_level=settings.ton_liteclient_trust_level,
            timeout_seconds=settings.ton_liteclient_timeout_seconds,
        )
    except TonLiteclientJettonVerificationFailure as exc:
        raise WalletJettonContractVerificationFailure(str(exc)) from exc
    _validate_live_result(
        verified,
        expected_trust_level=settings.ton_liteclient_trust_level,
    )
    verified_at = datetime.now(timezone.utc)
    values = {
        "run_id": run_id,
        "balance_snapshot_id": snapshot.id,
        "contract_version": JETTON_CONTRACT_VERIFICATION_VERSION,
        "verifier_name": verified["verifier_name"],
        "verifier_version": verified["verifier_version"],
        "network": network,
        "trust_level": verified["trust_level"],
        "anchor_workchain": verified["anchor"]["workchain"],
        "anchor_shard": verified["anchor"]["shard"],
        "anchor_seqno": verified["anchor"]["seqno"],
        "anchor_root_hash": verified["anchor"]["root_hash"],
        "anchor_file_hash": verified["anchor"]["file_hash"],
        "owner_account_canonical": run.wallet_address_canonical,
        "jetton_wallet_account_canonical": wallet,
        "jetton_master_account_canonical": master,
        "asset_identity_key": f"{JETTON_ASSET_IDENTITY_VERSION}|{network}|{master}",
        "wallet_balance_base_units": verified["wallet_balance_base_units"],
        "total_supply_base_units": verified["total_supply_base_units"],
        "mintable": verified["mintable"],
        "wallet_code_boc_hex": verified["wallet_code_boc_hex"],
        "wallet_data_boc_hex": verified["wallet_data_boc_hex"],
        "master_code_boc_hex": verified["master_code_boc_hex"],
        "master_data_boc_hex": verified["master_data_boc_hex"],
        "wallet_code_hash": verified["wallet_code_hash"],
        "wallet_data_hash": verified["wallet_data_hash"],
        "master_code_hash": verified["master_code_hash"],
        "master_data_hash": verified["master_data_hash"],
        "jetton_content_hash": verified["jetton_content_hash"],
        "verified_at": verified_at,
    }
    values["evidence_digest_sha256"] = _digest_json(
        _evidence_document(values)
    )
    record = WalletJettonContractVerification(**values)
    session.add(record)
    session.flush()
    result = _record_response(record)
    session.commit()
    return result


def list_wallet_jetton_contract_verifications(
    run_id: int,
    session: Session,
) -> dict[str, Any]:
    run = session.get(WalletIngestionRun, run_id)
    if run is None:
        raise WalletJettonContractVerificationNotFound(
            "Persisted wallet ingestion run not found."
        )
    network = _validated_run_network(run)
    records = list(
        session.scalars(
            select(WalletJettonContractVerification)
            .where(WalletJettonContractVerification.run_id == run_id)
            .order_by(WalletJettonContractVerification.id)
        )
    )
    items = [_record_response(record) for record in records]
    document = {
        "contract_version": JETTON_CONTRACT_VERIFICATION_VERSION,
        "run_id": str(run_id),
        "network": network,
        "verification_count": len(items),
        "verification_digests": [
            item["evidence_digest_sha256"] for item in items
        ],
    }
    return {
        **document,
        "verifications": items,
        "catalog_digest_sha256": _digest_json(document),
        "provider_requests_performed": False,
        "raw_account_state_bocs_returned": False,
        "message": (
            "Stored jetton contract verifications were reparsed and digest-"
            "checked provider-free. Raw account state BOCs remain hidden."
        ),
    }


def _record_response(record: WalletJettonContractVerification) -> dict[str, Any]:
    values = {column.name: getattr(record, column.name) for column in record.__table__.columns}
    for field, hash_field in (
        ("wallet_code_boc_hex", "wallet_code_hash"),
        ("wallet_data_boc_hex", "wallet_data_hash"),
        ("master_code_boc_hex", "master_code_hash"),
        ("master_data_boc_hex", "master_data_hash"),
    ):
        cell = _one_boc(values[field], field)
        if cell.hash.hex() != values[hash_field]:
            raise WalletJettonContractVerificationConflict(
                "Stored jetton account-state BOC hash changed."
            )
    document = _evidence_document(values)
    digest = _digest_json(document)
    if digest != record.evidence_digest_sha256:
        raise WalletJettonContractVerificationConflict(
            "Stored jetton contract verification digest changed."
        )
    return {
        "contract_version": record.contract_version,
        "verification_id": str(record.id),
        "run_id": str(record.run_id),
        "balance_snapshot_id": str(record.balance_snapshot_id),
        "verifier_name": record.verifier_name,
        "verifier_version": record.verifier_version,
        "network": record.network,
        "trust_level": record.trust_level,
        "anchor": document["anchor"],
        "owner_account_canonical": record.owner_account_canonical,
        "jetton_wallet_account_canonical": record.jetton_wallet_account_canonical,
        "jetton_master_account_canonical": record.jetton_master_account_canonical,
        "asset_identity_key": record.asset_identity_key,
        "wallet_balance_base_units": record.wallet_balance_base_units,
        "total_supply_base_units": record.total_supply_base_units,
        "mintable": record.mintable,
        "wallet_code_hash": record.wallet_code_hash,
        "wallet_data_hash": record.wallet_data_hash,
        "master_code_hash": record.master_code_hash,
        "master_data_hash": record.master_data_hash,
        "jetton_content_hash": record.jetton_content_hash,
        "account_state_boc_hashes": document["account_state_boc_hashes"],
        "evidence_digest_sha256": digest,
        "verified_at": _iso(record.verified_at),
        "account_state_proof_verified": True,
        "masterchain_checkpoint_chain_verified": record.trust_level == 0,
        "local_tvm_execution_applied": True,
        "wallet_owner_master_verified": True,
        "master_wallet_address_verified": True,
        "wallet_code_consistency_verified": True,
        "jetton_asset_identity_applied": True,
        "raw_account_state_bocs_persisted": True,
        "raw_account_state_bocs_returned": False,
        "is_blockchain_inclusion_proof_verified": record.trust_level == 0,
        "eligible_for_cost_basis": False,
        "used_by_pnl": False,
        "is_ownership_proof": False,
        "message": (
            "Proof-checked account states and locally executed jetton getters "
            "agree on owner, master, wallet address, and wallet code. This "
            "establishes a network-scoped jetton asset relationship, not wallet "
            "ownership, complete activity history, cost basis, or PnL."
        ),
    }


def _evidence_document(values: dict[str, Any]) -> dict[str, Any]:
    return {
        "contract_version": values["contract_version"],
        "run_id": values["run_id"],
        "balance_snapshot_id": values["balance_snapshot_id"],
        "verifier_name": values["verifier_name"],
        "verifier_version": values["verifier_version"],
        "network": values["network"],
        "trust_level": values["trust_level"],
        "anchor": {
            "workchain": values["anchor_workchain"],
            "shard": values["anchor_shard"],
            "seqno": values["anchor_seqno"],
            "root_hash": values["anchor_root_hash"],
            "file_hash": values["anchor_file_hash"],
        },
        "owner_account_canonical": values["owner_account_canonical"],
        "jetton_wallet_account_canonical": values[
            "jetton_wallet_account_canonical"
        ],
        "jetton_master_account_canonical": values[
            "jetton_master_account_canonical"
        ],
        "asset_identity_key": values["asset_identity_key"],
        "wallet_balance_base_units": values["wallet_balance_base_units"],
        "total_supply_base_units": values["total_supply_base_units"],
        "mintable": values["mintable"],
        "wallet_code_hash": values["wallet_code_hash"],
        "wallet_data_hash": values["wallet_data_hash"],
        "master_code_hash": values["master_code_hash"],
        "master_data_hash": values["master_data_hash"],
        "jetton_content_hash": values["jetton_content_hash"],
        "account_state_boc_hashes": {
            field: hashlib.sha256(bytes.fromhex(values[field])).hexdigest()
            for field in (
                "wallet_code_boc_hex",
                "wallet_data_boc_hex",
                "master_code_boc_hex",
                "master_data_boc_hex",
            )
        },
        "verified_at": _iso(values["verified_at"]),
    }


def _exact_snapshot(
    run_id: int,
    wallet: str,
    master: str,
    network: str,
    session: Session,
) -> WalletBalanceSnapshot:
    matches = []
    snapshots = session.scalars(
        select(WalletBalanceSnapshot).where(
            WalletBalanceSnapshot.run_id == run_id,
            WalletBalanceSnapshot.provider == "tonapi",
            WalletBalanceSnapshot.source_status == "live",
        )
    )
    for snapshot in snapshots:
        try:
            raw = json.loads(snapshot.raw_json or "")
        except (json.JSONDecodeError, TypeError):
            continue
        if not isinstance(raw, dict) or raw.get("surface") != "jettons":
            continue
        raw_wallet = _canonical_account(raw.get("wallet_contract_address"), network)
        raw_master = _canonical_account(raw.get("jetton_address"), network)
        if raw_wallet == wallet and raw_master == master:
            matches.append(snapshot)
    if not matches:
        raise WalletJettonContractVerificationNotFound(
            "Exact persisted live TonAPI jetton snapshot relation not found."
        )
    if len(matches) != 1:
        raise WalletJettonContractVerificationConflict(
            "Multiple persisted snapshots claim the same run-scoped jetton relation."
        )
    return matches[0]


def _validated_run_network(run: WalletIngestionRun) -> str:
    if (
        run.data_mode != "real"
        or run.wallet_identity_status != "network_scoped"
        or run.wallet_network not in {"ton-mainnet", "ton-testnet"}
        or run.wallet_address_canonical is None
    ):
        raise WalletJettonContractVerificationConflict(
            "Jetton contract verification requires a real network-scoped run."
        )
    return run.wallet_network


def _canonical_account(value: Any, network: str) -> str | None:
    identity = derive_ton_wallet_identity(value, network_context=network)
    if identity.status != "network_scoped" or identity.network != network:
        return None
    return identity.canonical_address


def _validate_live_result(
    value: dict[str, Any],
    *,
    expected_trust_level: int,
) -> None:
    required = {
        "verifier_name",
        "verifier_version",
        "trust_level",
        "anchor",
        "wallet_balance_base_units",
        "total_supply_base_units",
        "mintable",
        "wallet_code_boc_hex",
        "wallet_data_boc_hex",
        "master_code_boc_hex",
        "master_data_boc_hex",
        "wallet_code_hash",
        "wallet_data_hash",
        "master_code_hash",
        "master_data_hash",
        "jetton_content_hash",
        "account_state_proof_verified",
        "masterchain_checkpoint_chain_verified",
        "local_tvm_execution_applied",
    }
    if not isinstance(value, dict) or set(value) != required:
        raise WalletJettonContractVerificationConflict(
            "Local jetton verifier returned an unexpected result contract."
        )
    anchor = value["anchor"]
    canonical_integer = re.compile(r"^(?:0|[1-9][0-9]*)$")
    canonical_hash = re.compile(r"^[0-9a-f]{64}$")
    canonical_shard = re.compile(r"^-?[0-9]{1,20}$")
    if not (
        value["verifier_name"] == "pytoniq-pytvm"
        and isinstance(value["verifier_version"], str)
        and 1 <= len(value["verifier_version"]) <= 48
        and type(value["trust_level"]) is int
        and value["trust_level"] == expected_trust_level
        and isinstance(anchor, dict)
        and set(anchor)
        == {"workchain", "shard", "seqno", "root_hash", "file_hash"}
        and type(anchor["workchain"]) is int
        and anchor["workchain"] in {-1, 0}
        and isinstance(anchor["shard"], str)
        and canonical_shard.fullmatch(anchor["shard"])
        and type(anchor["seqno"]) is int
        and anchor["seqno"] > 0
        and isinstance(anchor["root_hash"], str)
        and canonical_hash.fullmatch(anchor["root_hash"])
        and isinstance(anchor["file_hash"], str)
        and canonical_hash.fullmatch(anchor["file_hash"])
        and isinstance(value["wallet_balance_base_units"], str)
        and len(value["wallet_balance_base_units"]) <= 80
        and canonical_integer.fullmatch(value["wallet_balance_base_units"])
        and isinstance(value["total_supply_base_units"], str)
        and len(value["total_supply_base_units"]) <= 80
        and canonical_integer.fullmatch(value["total_supply_base_units"])
        and type(value["mintable"]) is bool
        and isinstance(value["jetton_content_hash"], str)
        and canonical_hash.fullmatch(value["jetton_content_hash"])
    ):
        raise WalletJettonContractVerificationConflict(
            "Local jetton verifier returned malformed bounded evidence."
        )
    if not (
        value["account_state_proof_verified"] is True
        and value["local_tvm_execution_applied"] is True
        and value["masterchain_checkpoint_chain_verified"]
        == (value["trust_level"] == 0)
    ):
        raise WalletJettonContractVerificationConflict(
            "Local jetton verifier proof flags are incoherent."
        )
    for field, hash_field in (
        ("wallet_code_boc_hex", "wallet_code_hash"),
        ("wallet_data_boc_hex", "wallet_data_hash"),
        ("master_code_boc_hex", "master_code_hash"),
        ("master_data_boc_hex", "master_data_hash"),
    ):
        if not (
            isinstance(value[hash_field], str)
            and canonical_hash.fullmatch(value[hash_field])
            and _one_boc(value[field], field).hash.hex() == value[hash_field]
        ):
            raise WalletJettonContractVerificationConflict(
                "Local jetton verifier BOC/hash evidence is incoherent."
            )


def _one_boc(value: Any, field: str) -> Cell:
    if (
        not isinstance(value, str)
        or len(value) > 2 * 1024 * 1024
        or len(value) % 2
        or value != value.lower()
    ):
        raise WalletJettonContractVerificationConflict(
            f"Stored {field} is not a bounded canonical BOC."
        )
    try:
        roots = Cell.from_boc(bytes.fromhex(value))
    except Exception as exc:
        raise WalletJettonContractVerificationConflict(
            f"Stored {field} could not be reparsed."
        ) from exc
    if len(roots) != 1:
        raise WalletJettonContractVerificationConflict(
            f"Stored {field} must contain exactly one BOC root."
        )
    return roots[0]


def _digest_json(value: Any) -> str:
    return hashlib.sha256(
        json.dumps(
            value,
            ensure_ascii=True,
            sort_keys=True,
            separators=(",", ":"),
            allow_nan=False,
        ).encode("utf-8")
    ).hexdigest()


def _iso(value: datetime) -> str:
    if value.tzinfo is None:
        value = value.replace(tzinfo=timezone.utc)
    return value.astimezone(timezone.utc).isoformat().replace("+00:00", "Z")


__all__ = [
    "JETTON_ASSET_IDENTITY_VERSION",
    "JETTON_CONTRACT_VERIFICATION_VERSION",
    "WalletJettonContractVerificationConflict",
    "WalletJettonContractVerificationFailure",
    "WalletJettonContractVerificationNotFound",
    "list_wallet_jetton_contract_verifications",
    "verify_wallet_jetton_contract_relationship",
]

File: frontend/src/walletJettonContractVerification.ts
import { parseRfc3339Instant } from "./rfc3339";
import type {
  WalletAccountStateInclusionProofRecord,
  WalletJettonContractVerificationAnchorRecord,
  WalletJettonContractVerificationCatalogResponse,
  WalletJettonContractVerificationResponse,
} from "./types";

const HASH = /^[0-9a-f]{64}$/;
const ADDRESS = /^(?:-1|0):[0-9a-f]{64}$/;
const POSITIVE_ID = /^[1-9][0-9]{0,18}$/;
const BASE_UNITS = /^(?:0|[1-9][0-9]*)$/;
const MIN_INT64 = -9_223_372_036_854_775_808n;
const MAX_INT64 = 9_223_372_036_854_775_807n;
const CHECKPOINT_LIMITATION = "checkpoint_policy_not_persisted_v1" as const;

const RESPONSE_KEYS = [
  "account_state_boc_hashes",
  "account_state_inclusion_proofs",
  "account_state_proof_verified",
  "anchor",
  "asset_identity_key",
  "balance_snapshot_id",
  "contract_version",
  "eligible_for_cost_basis",
  "evidence_digest_sha256",
  "is_blockchain_inclusion_proof_verified",
  "is_ownership_proof",
  "jetton_asset_identity_applied",
  "jetton_content_hash",
  "jetton_master_account_canonical",
  "jetton_wallet_account_canonical",
  "limitations",
  "local_tvm_execution_applied",
  "master_code_hash",
  "master_data_hash",
  "master_wallet_address_verified",
  "masterchain_checkpoint_chain_verified",
  "message",
  "mintable",
  "network",
  "owner_account_canonical",
  "raw_account_state_bocs_persisted",
  "raw_account_state_bocs_returned",
  "run_id",
  "total_supply_base_units",
  "trust_level",
  "used_by_pnl",
  "verification_id",
  "verified_at",
  "verifier_name",
  "verifier_version",
  "wallet_balance_base_units",
  "wallet_code_consistency_verified",
  "wallet_code_hash",
  "wallet_data_hash",
  "wallet_owner_master_verified",
] as const;
const CATALOG_KEYS = [
  "catalog_digest_sha256",
  "contract_version",
  "message",
  "network",
  "provider_requests_performed",
  "raw_account_state_bocs_returned",
  "run_id",
  "verification_count",
  "verification_digests",
  "verifications",
] as const;
const ANCHOR_KEYS = ["file_hash", "root_hash", "seqno", "shard", "workchain"] as const;
const ACCOUNT_BOC_HASH_KEYS = [
  "master_code_boc_hex",
  "master_data_boc_hex",
  "wallet_code_boc_hex",
  "wallet_data_boc_hex",
] as const;
const PROOF_KEYS = [
  "account_address_canonical",
  "account_role",
  "boc_sha256",
  "contract_version",
  "evidence_digest_sha256",
  "provider_free_revalidated",
  "provider_requests_performed",
  "raw_bocs_returned",
  "shard_block",
  "verified_at",
] as const;
const PROOF_BOC_HASH_KEYS = [
  "account_proof_boc_hex",
  "shard_proof_boc_hex",
  "state_boc_hex",
] as const;

function fail(): never {
  throw new Error("Jetton contract proof response is incoherent.");
}

export function validateWalletJettonContractVerification(
  value: unknown,
  expectedRunId: number,
  expectedNetwork: "ton-mainnet" | "ton-testnet",
): WalletJettonContractVerificationResponse {
  if (!isRecord(value) || !hasExactKeys(value, RESPONSE_KEYS)) fail();
  const row = value;
  if (
    row.contract_version !== "ton_jetton_contract_verification_v1" ||
    row.run_id !== String(expectedRunId) ||
    row.network !== expectedNetwork ||
    !isPositiveId(row.verification_id) ||
    !isPositiveId(row.run_id) ||
    !isPositiveId(row.balance_snapshot_id) ||
    row.verifier_name !== "pytoniq-pytvm" ||
    typeof row.verifier_version !== "string" ||
    row.verifier_version.length < 1 ||
    row.verifier_version.length > 48 ||
    (row.trust_level !== 0 && row.trust_level !== 1) ||
    !validAnchor(row.anchor) ||
    !isAddress(row.owner_account_canonical) ||
    !isAddress(row.jetton_wallet_account_canonical) ||
    !isAddress(row.jetton_master_account_canonical) ||
    row.jetton_wallet_account_canonical === row.jetton_master_account_canonical ||
    row.asset_identity_key !==
      `ton_jetton_asset_v1|${expectedNetwork}|${row.jetton_master_account_canonical}` ||
    !isBaseUnits(row.wallet_balance_base_units) ||
    !isBaseUnits(row.total_supply_base_units) ||
    typeof row.mintable !== "boolean" ||
    !isHash(row.wallet_code_hash) ||
    !isHash(row.wallet_data_hash) ||
    !isHash(row.master_code_hash) ||
    !isHash(row.master_data_hash) ||
    !isHash(row.jetton_content_hash) ||
    !validHashRecord(row.account_state_boc_hashes, ACCOUNT_BOC_HASH_KEYS) ||
    !isHash(row.evidence_digest_sha256) ||
    !isUtcTimestamp(row.verified_at) ||
    row.account_state_proof_verified !== true ||
    row.masterchain_checkpoint_chain_verified !== false ||
    row.local_tvm_execution_applied !== true ||
    row.wallet_owner_master_verified !== true ||
    row.master_wallet_address_verified !== true ||
    row.wallet_code_consistency_verified !== true ||
    row.jetton_asset_identity_applied !== true ||
    typeof row.raw_account_state_bocs_persisted !== "boolean" ||
    row.raw_account_state_bocs_returned !== false ||
    row.is_blockchain_inclusion_proof_verified !== false ||
    row.eligible_for_cost_basis !== false ||
    row.used_by_pnl !== false ||
    row.is_ownership_proof !== false ||
    !Array.isArray(row.limitations) ||
    row.limitations.length !== 1 ||
    row.limitations[0] !== CHECKPOINT_LIMITATION ||
    typeof row.message !== "string" ||
    row.message.length < 1 ||
    row.message.length > 600 ||
    !Array.isArray(row.account_state_inclusion_proofs) ||
    (row.account_state_inclusion_proofs.length !== 0 &&
      row.account_state_inclusion_proofs.length !== 2) ||
    row.raw_account_state_bocs_persisted !==
      (row.account_state_inclusion_proofs.length === 2)
  ) {
    fail();
  }

  const expectedRoles = ["jetton_master", "jetton_wallet"] as const;
  row.account_state_inclusion_proofs.forEach((proof, index) => {
    const role = expectedRoles[index];
    const address = role === "jetton_master"
      ? row.jetton_master_account_canonical
      : row.jetton_wallet_account_canonical;
    if (!validAccountInclusionProof(proof, role, address, row.verified_at)) fail();
  });

  return value as unknown as WalletJettonContractVerificationResponse;
}

export function validateWalletJettonContractVerificationCatalog(
  value: unknown,
  expectedRunId: number,
  expectedNetwork: "ton-mainnet" | "ton-testnet",
): WalletJettonContractVerificationCatalogResponse {
  if (!isRecord(value) || !hasExactKeys(value, CATALOG_KEYS)) fail();
  const catalog = value;
  const verificationDigests = catalog.verification_digests;
  const verifications = catalog.verifications;
  if (
    catalog.contract_version !== "ton_jetton_contract_verification_v1" ||
    catalog.run_id !== String(expectedRunId) ||
    !isPositiveId(catalog.run_id) ||
    catalog.network !== expectedNetwork ||
    !Number.isSafeInteger(catalog.verification_count) ||
    (catalog.verification_count as number) < 0 ||
    (catalog.verification_count as number) > 500 ||
    !Array.isArray(verificationDigests) ||
    !Array.isArray(verifications) ||
    catalog.provider_requests_performed !== false ||
    catalog.raw_account_state_bocs_returned !== false ||
    !isHash(catalog.catalog_digest_sha256) ||
    typeof catalog.message !== "string" ||
    catalog.message.length < 1 ||
    catalog.message.length > 400 ||
    catalog.verification_count !== verifications.length ||
    verificationDigests.length !== verifications.length ||
    verificationDigests.some((digest) => !isHash(digest))
  ) {
    fail();
  }
  verifications.forEach((proof, index) => {
    const validated = validateWalletJettonContractVerification(
      proof,
      expectedRunId,
      expectedNetwork,
    );
    if (verificationDigests[index] !== validated.evidence_digest_sha256) fail();
  });
  return value as unknown as WalletJettonContractVerificationCatalogResponse;
}

function validAccountInclusionProof(
  value: unknown,
  role: "jetton_master" | "jetton_wallet",
  address: unknown,
  verifiedAt: unknown,
): value is WalletAccountStateInclusionProofRecord {
  if (!isRecord(value) || !hasExactKeys(value, PROOF_KEYS)) return false;
  return (
    value.contract_version === "ton_account_state_inclusion_v1" &&
    value.account_role === role &&
    value.account_address_canonical === address &&
    isAddress(value.account_address_canonical) &&
    validAnchor(value.shard_block) &&
    validHashRecord(value.boc_sha256, PROOF_BOC_HASH_KEYS) &&
    isHash(value.evidence_digest_sha256) &&
    value.verified_at === verifiedAt &&
    isUtcTimestamp(value.verified_at) &&
    value.provider_requests_performed === false &&
    value.provider_free_revalidated === true &&
    value.raw_bocs_returned === false
  );
}

function validAnchor(value: unknown): value is WalletJettonContractVerificationAnchorRecord {
  if (!isRecord(value) || !hasExactKeys(value, ANCHOR_KEYS)) return false;
  return (
    (value.workchain === -1 || value.workchain === 0) &&
    isCanonicalInt64(value.shard) &&
    Number.isSafeInteger(value.seqno) &&
    (value.seqno as number) > 0 &&
    (value.seqno as number) <= 2_147_483_647 &&
    isHash(value.root_hash) &&
    isHash(value.file_hash)
  );
}

function validHashRecord(value: unknown, keys: readonly string[]): boolean {
  return (
    isRecord(value) &&
    hasExactKeys(value, keys) &&
    Object.values(value).every(isHash)
  );
}

function isPositiveId(value: unknown): value is string {
  return typeof value === "string" && POSITIVE_ID.test(value);
}

function isAddress(value: unknown): value is string {
  return typeof value === "string" && ADDRESS.test(value);
}

function isBaseUnits(value: unknown): value is string {
  return typeof value === "string" && BASE_UNITS.test(value);
}

function isHash(value: unknown): value is string {
  return typeof value === "string" && HASH.test(value);
}

function isUtcTimestamp(value: unknown): value is string {
  return (
    typeof value === "string" &&
    value.length <= 40 &&
    parseRfc3339Instant(value, { requireUtc: true, maximumFractionDigits: 6 }) !== null
  );
}

function isCanonicalInt64(value: unknown): value is string {
  if (
    typeof value !== "string" ||
    !/^(?:0|[1-9][0-9]*|-[1-9][0-9]*)$/.test(value) ||
    value.length > 20
  ) return false;
  const parsed = BigInt(value);
  return parsed >= MIN_INT64 && parsed <= MAX_INT64;
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function hasExactKeys(value: Record<string, unknown>, expected: readonly string[]): boolean {
  const actual = Object.keys(value).sort();
  const sortedExpected = [...expected].sort();
  return (
    actual.length === sortedExpected.length &&
    actual.every((key, index) => key === sortedExpected[index])
  );
}
